src/main/java/Duke.java

```diff
@@ -1,10 +1,166 @@
+                bye();
+                exit = true;
+                break;
+            case "list":
+                list(tasks);
+                break;
+            case "done":
+                done(tasks, arguments);
+                break;
+            case "deadline":
+                deadline(tasks, arguments);
+                break;
+            case "event":
+                event(tasks, arguments);
+                break;
+            case "todo":
+                todo(tasks, arguments);
+                break;
+            default:
+                save(tasks, line);
+            }
+
+            printlnWithIndent(LONG_LINE);
+        }
+        in.close();
+    }
+
+    // Print a greeting message when the program is invoked
+    protected static void greeting() {
+        printlnWithIndent("Hello! I'm Duke.");
+        printlnWithIndent("What can I do for you?");
+    }
+
+    // Print a goodbye message before the program exits
+    protected static void bye() {
+        printlnWithIndent("Bye. Hope to see you again soon!");
+    }
+
+    // Print out everything in the list, index starts from 1
+    protected static void list(Vector<Task> tasks) {
+        printlnWithIndent("Here are the tasks in your list:");
+        for (int i = 0; i < tasks.size(); i += 1) {
+            printlnWithIndent(String.format("%d.\t%s", i + 1, tasks.get(i)));
+        }
+    }
+
+    // Mark a task to be done with index specified in arguments[1]
+    protected static void done(Vector<Task> tasks, String[] arguments) {
+        if (arguments.length < 2) {
+            // An index must be provided for the task to be marked "done"
+            printlnWithIndent("You will need to give me an index, like this: `done 2`.");
+        } else {
+            try {
+                int index = Integer.parseInt(arguments[1]);
+                if (index > tasks.size() || index < 1) {
+                    // This index is out of the boundary of our database
+                    throw new IllegalArgumentException();
+                }
+
+                Task task = tasks.get(index - 1);
+                task.markAsDone();
+                tasks.set(index - 1, task);
+
+                printlnWithIndent("Nice! I've marked this task as done:");
+                printlnWithIndent("\t" + task);
+            } catch (NumberFormatException e) {
+                printlnWithIndent("Index provided is not a proper number.");
+            } catch (IllegalArgumentException e) {
+                printlnWithIndent("Task with this index is not found in our database.");
+            }
+        }
+    }
+
+    // Create a deadline task
+    protected static void deadline(Vector<Task> tasks, String[] arguments) {
+        int i = findIndex(arguments, "/by");
+        if (i != -1) {
+            String description = String.join(" ", Arrays.copyOfRange(arguments, 1, i));
+            String by = String.join(" ", Arrays.copyOfRange(arguments, i + 1, arguments.length));
```

Comment:
Is the use of String.join() necessary here? Unless I'm missing something, it is possible to achieve the same result by just indexing the arguments array

Reply:
Thank you for the advice. I wanted to use `String.join` and `Arrays.copyOfRange` to split the tokenized arguments (and also remove the 1st element of the tokens as that's a command like `todo`, `deadline` etc.). May I know if there is a better solution for this?

You mentioned by just indexing the arguments array, do you mean something like this:
```
String by = "";
for (int j = i + 1; j < arguments.length; ++j) {
    by += arguments[j] + " ";
}
by = by.trim();
```